In [1]:
import calitp
from calitp.tables import tbl
from siuba import *

import pandas as pd
import numpy as np
import geopandas as gpd
import fiona

from ipyleaflet import Map, GeoJSON, projections, basemaps, GeoData, LayersControl, WidgetControl, GeoJSON
from ipywidgets import Text, HTML

import shapely
from shapely.geometry import LineString, MultiPoint
from shapely.ops import split, substring

import zlib
import datetime as dt

from utilities import *
import shared_utils

/opt/conda/lib/python3.9/site-packages/geopandas/_compat.py:111: UserWarning: The Shapely GEOS version (3.9.1-CAPI-1.14.2) is incompatible with the GEOS version PyGEOS was compiled with (3.10.1-CAPI-1.16.0). Conversions between both will be slow.
  warnings.warn(


### High Quality Transit Areas Relevant Statutes

[PRC 21155](https://leginfo.legislature.ca.gov/faces/codes_displaySection.xhtml?sectionNum=21155.&lawCode=PRC)
* Major transit stop definition: _A major transit stop is as defined in Section 21064.3, except that, for purposes of this section, it also includes major transit stops that are included in the applicable regional transportation plan_
* High-quality transit corridor definition: _For purposes of this section, a high-quality transit corridor means a corridor with fixed route bus service with service intervals no longer than 15 minutes during peak commute hours._
    * Unable to locate definition of "peak commute hours"

[PRC 21064.3](https://leginfo.legislature.ca.gov/faces/codes_displaySection.xhtml?sectionNum=21064.3.&lawCode=PRC)
* _Major transit stop means a site containing any of the following:
(a) An existing rail or bus rapid transit station.
(b) A ferry terminal served by either a bus or rail transit service.
(c) The intersection of two or more major bus routes with a frequency of service interval of 15 minutes or less during the morning and afternoon peak commute periods._
    * "Intersection" may not be sufficiently well-defined for this analysis

[PRC 21060.2](https://leginfo.legislature.ca.gov/faces/codes_displaySection.xhtml?lawCode=PRC&sectionNum=21060.2.&highlight=true&keyword=bus%20rapid%20transit)
* _(a) “Bus rapid transit” means a public mass transit service provided by a public agency or by a public-private partnership that includes all of the following features:
(1) Full-time dedicated bus lanes or operation in a separate right-of-way dedicated for public transportation with a frequency of service interval of 15 minutes or less during the morning and afternoon peak commute periods.
(2) Transit signal priority.
(3) All-door boarding.
(4) Fare collection system that promotes efficiency.
(5) Defined stations._
    * Unlikely to determine if a service qualifies as BRT under this definition using GTFS alone

In [2]:
# ##definition of "peak" pending
# am_peak = range(6, 9)
# pm_peak = range(16, 19)

In [3]:
# am_commute_hours = list(am_peak)
# pm_commute_hours = list(pm_peak)
# commute_hours = am_commute_hours + pm_commute_hours

In [4]:
def create_segments(geometry):
    '''Splits a Shapely LineString into smaller LineStrings. If a MultiLineString passed,
    splits each LineString in that collection. 
    '''
    
    lines = []
    segment_distance_meters = 1250
    geometry = geometry.iloc[0]
    if hasattr(geometry, 'geoms'): ##check if MultiLineString
        linestrings = geometry.geoms
    else:
        linestrings = [geometry]
    for linestring in linestrings:
        for i in range(0, int(linestring.length), segment_distance_meters):
            lines.append(substring(linestring, i, i+segment_distance_meters))
    return lines

In [5]:
def find_stop_with_high_trip_count(segment, stops, stop_times, rank, calculated_stops):
    '''Given a shape segment, finds the stop serving the most (or other rank) trips within that segment.
    Adds that stop's stop_id to segment data (a row).
    '''
    
    stops_in_seg = gpd.clip(stops, segment.geometry)
    if stops_in_seg.size == 0:
        return segment
    
    stop_times_in_seg = stops_in_seg >> inner_join(_, stop_times, on = 'stop_id')
    trip_count_by_stop = stop_times_in_seg >> count(_.stop_id) >> arrange(-_.n) >> rename(n_trips = _.n)
    try:
        stop_id = trip_count_by_stop['stop_id'].iloc[rank - 1]

        if stop_id in list(calculated_stops):
            return segment
        segment['stop_id'] = stop_id
        segment['n_trips'] = trip_count_by_stop['n_trips'].iloc[rank - 1]
        return segment
    except IndexError:
        return segment

In [6]:
# itp_id = 170

# most_recent_wed = (tbl.views.dim_date()
#  >> filter(_.full_date < dt.datetime.now().date())
#  >> filter(_.day_name == 'Wednesday')
#  >> filter(_.full_date == _.full_date.max())
#  >> select(_.service_date == _.full_date)
# )
# wednesday = (tbl.views.gtfs_schedule_fact_daily_service() 
#          >> filter(_.calitp_itp_id == int(itp_id))
#          >> inner_join(_, most_recent_wed, on = 'service_date')
#          >> collect()
#             )
# wednesday = wednesday >> select(_.calitp_itp_id, _.calitp_url_number, _.service_id)

In [7]:
# bus_routes = (tbl.gtfs_schedule.routes()
#             >> filter(_.calitp_itp_id == int(itp_id))
#             >> filter((_.route_type == '3') | (_.route_type == '11'))
#             >> select(_.calitp_itp_id, _.calitp_url_number, _.route_id) 
#             >> collect()
#          )

In [8]:
# trips = (tbl.gtfs_schedule.trips()
#          >> filter(_.calitp_itp_id == int(itp_id))
#          >> collect()
#          >> inner_join(_, bus_routes, on = ['calitp_itp_id', 'calitp_url_number', 'route_id'])
#          >> inner_join(_, wednesday, on = ['calitp_itp_id', 'calitp_url_number', 'service_id'])
#          )

In [9]:
# stop_times = (tbl.gtfs_schedule.stop_times()
#               >> filter(_.calitp_itp_id == int(itp_id))
#               >> collect()
#              )

In [10]:
# stop_times =  (stop_times >> inner_join(_, trips, on = ['calitp_itp_id', 'calitp_url_number', 'trip_id'])
#               >> select(-_.stop_headsign, -_.pickup_type, -_.drop_off_type,
#                         -_.continuous_pickup, -_.continuous_drop_off, -_.shape_dist_travelled,
#                        -_.timepoint)
#               )

In [11]:
# stops = (tbl.gtfs_schedule.stops() 
#          >> filter(_.calitp_itp_id == itp_id)
#          >> select(_.stop_id, _.stop_lat, _.stop_lon)
#          >> collect())
# stops = gpd.GeoDataFrame(stops,
#                  geometry = gpd.points_from_xy(stops.stop_lon, stops.stop_lat),
#                  crs = 'EPSG:4326').to_crs(shared_utils.geography_utils.CA_NAD83Albers)

In [12]:
def get_operator_views(itp_id):
    '''Returns relevant views from the data warehouse for a single transit operator.
    '''
    shapes = tbl.gtfs_schedule.shapes() >> filter(_.calitp_itp_id == int(itp_id)) >> collect()
    shapes = gpd.GeoDataFrame(shapes, 
                              geometry = gpd.points_from_xy(shapes.shape_pt_lon, shapes.shape_pt_lat),
                              crs = 'EPSG:4326').to_crs(shared_utils.geography_utils.CA_NAD83Albers)
    most_recent_wed = (tbl.views.dim_date()
     >> filter(_.full_date < dt.datetime.now().date())
     >> filter(_.day_name == 'Wednesday')
     >> filter(_.full_date == _.full_date.max())
     >> select(_.service_date == _.full_date)
    )
    wednesday = (tbl.views.gtfs_schedule_fact_daily_service() 
             >> filter(_.calitp_itp_id == int(itp_id))
             >> inner_join(_, most_recent_wed, on = 'service_date')
             >> collect()
                )
    wednesday = wednesday >> select(_.calitp_itp_id, _.calitp_url_number, _.service_id)

    bus_routes = (tbl.gtfs_schedule.routes()
                >> filter(_.calitp_itp_id == int(itp_id))
                >> filter((_.route_type == '3') | (_.route_type == '11'))
                >> select(_.calitp_itp_id, _.calitp_url_number, _.route_id) 
                >> collect()
             )
    print('loaded bus routes')
    
    trips = (tbl.gtfs_schedule.trips()
             >> filter(_.calitp_itp_id == int(itp_id))
             >> collect()
             >> inner_join(_, bus_routes, on = ['calitp_itp_id', 'calitp_url_number', 'route_id'])
             >> inner_join(_, wednesday, on = ['calitp_itp_id', 'calitp_url_number', 'service_id'])
             )
    print('loaded trips')
    stop_times = (tbl.gtfs_schedule.stop_times()
                  >> filter(_.calitp_itp_id == int(itp_id))
                  >> collect()
                 )
    stop_times =  (stop_times >> inner_join(_, trips, on = ['calitp_itp_id', 'calitp_url_number', 'trip_id'])
                  >> select(-_.stop_headsign, -_.pickup_type, -_.drop_off_type,
                            -_.continuous_pickup, -_.continuous_drop_off, -_.shape_dist_travelled,
                           -_.timepoint)
                  )
    print('loaded stop times')

    stops = (tbl.gtfs_schedule.stops() 
             >> filter(_.calitp_itp_id == itp_id)
             >> select(_.stop_id, _.stop_lat, _.stop_lon)
             >> collect())
    stops = gpd.GeoDataFrame(stops,
                     geometry = gpd.points_from_xy(stops.stop_lon, stops.stop_lat),
                     crs = 'EPSG:4326').to_crs(shared_utils.geography_utils.CA_NAD83Albers)
    print('loaded stops')

    return shapes, trips, stop_times, stops

In [13]:
def fix_arrival_time(gtfs_timestring):
    '''Reformats a GTFS timestamp (which allows the hour to exceed 24 to mark service day continuity)
    to standard 24-hour time.
    '''
    split = gtfs_timestring.split(':')
    hour = int(split[0])
    if hour >= 24:
        split[0] = str(hour - 24)
        corrected = (':').join(split)
        return corrected.strip()
    else:
        return gtfs_timestring.strip()

In [14]:
def single_shape_hqta(shapes, trips, stop_times, stops, route_count_by_stop, shape_id, hqta):
    '''Starting with a single GTFS shape, split that shape into segments and determine if each segment qualifies
    as an HQTA. Existing segments within a shape are dropped for that shape, since their peak frequency and 
    HQTA status would have already been calculated for a previous shape.
    '''
    
    single_shape = (shapes
         >> filter(_.shape_id == shape_id)
         >> mutate(shape_pt_sequence = _.shape_pt_sequence.astype('int64'))
         >> arrange(_.shape_pt_sequence) ##arrange, then convert to line to preserve order...
        )
    
    route_line = LineString(list(single_shape['geometry'])) ##TODO replace with shared util? (upstream in shapes...)
    single_line = single_shape[['calitp_itp_id', 'shape_id', 'calitp_extracted_at']].iloc[[0]] ##preserve info cols
    single_line['geometry'] = route_line
    single_line = gpd.GeoDataFrame(single_line, crs=shared_utils.geography_utils.CA_NAD83Albers)
    
    if shape_id in debug_ids:
        print(f'***debug shape*** {shape_id}')
        debug_dict[f'{shape_id}_single_line'] = single_line
        debug_dict[f'{shape_id}_hqta'] = hqta

    calculated_stops = []
    if hqta.size != 0:
        already_calculated = hqta.dissolve(by='calitp_itp_id') ## get single polygon of HQTA calculation complete area
        single_line = single_line.overlay(already_calculated, how='difference') ## drop calculation complete area from current shape
        if single_line.size == 0:
            segments_with_max_stop = None
            print(f'no line for shape {shape_id}')
            return

    segmented = pd.DataFrame()
    for segment in create_segments(single_line.geometry):
        to_append = single_line.drop(columns=['geometry'])
        to_append['geometry'] = segment
        segmented = segmented.append(to_append)
        
    segmented = segmented.reset_index()
    segmented['segment_sequence'] = segmented.index.astype(str)
    segmented = segmented.astype({'calitp_itp_id': str})
    
    ## compute (hopefully unique) hash of segment id that can be used across routes/operators
    segmented['hqta_segment_id'] = segmented.apply(lambda x:
                zlib.crc32((x.calitp_itp_id + x.shape_id + x.segment_sequence).encode('utf-8')), axis=1)
    
    segmented.geometry = segmented.buffer(50) ##generous buffer for street/sidewalk width? Required to spatially find stops within each segment

    if shape_id in debug_ids:
        debug_dict[f'{shape_id}_segmented'] = segmented
        debug_dict[f'{shape_id}_stops'] = stops
        debug_dict[f'{shape_id}_route_ct_by_stop'] = route_count_by_stop
     
    segments_with_max_stop = segmented.apply(find_stop_with_high_trip_count, axis=1,
                                             args = (stops, stop_times, 1, calculated_stops))
    
    if not 'stop_id' in segments_with_max_stop.columns:
        segments_with_top2_stops = None
        print(f'no stops for shape {shape_id}')
        return ## no stops within segment

    max_stop_times = (stop_times 
                 >> select(_.stop_id, _.trip_id, _.departure_time)
                 >> inner_join(_, segments_with_max_stop, on = "stop_id")
                ) ## filter stop_times to the key stops in each segment
    max_stop_times = max_stop_times.dropna(subset=['departure_time'])
    max_stop_times = max_stop_times.drop_duplicates(subset=['trip_id','hqta_segment_id']) ## filter duplicates for top2 approach
    
    if shape_id in debug_ids:
        debug_dict[f'{shape_id}_max_stop0'] = max_stop_times
    
    max_stop_times['departure_time'] = max_stop_times['departure_time'].apply(fix_arrival_time) ## reformat GTFS time to a format datetime can ingest
    max_stop_times['departure_dt'] = max_stop_times['departure_time'].apply(lambda x:
                                                                    dt.datetime.strptime(x, '%H:%M:%S'))
    max_stop_times['departure_hour'] = max_stop_times['departure_dt'].apply(lambda x: x.hour)
    
    # max_stop_times = max_stop_times >> filter(_.departure_hour.isin(commute_hours))
    
    if max_stop_times.size == 0:
        print(f'no commute hour trips for shape {shape_id}')
        return
    
    if shape_id in debug_ids:
        debug_dict[f'{shape_id}_max_stop'] = max_stop_times

    ## new flexible peak
    segment_am_max = (max_stop_times
                  >> count(_.hqta_segment_id, _.departure_hour)
                  >> filter(_.departure_hour < 12)
                  >> group_by(_.hqta_segment_id)
                  >> summarize(am_max_trips = _.n.max())
                 )
    
    segment_pm_max = (max_stop_times
                  >> count(_.hqta_segment_id, _.departure_hour)
                  >> filter(_.departure_hour >= 12)
                  >> group_by(_.hqta_segment_id)
                  >> summarize(pm_max_trips = _.n.max())
                 )
    try:
        segment_peak_service = (segment_am_max >> inner_join(_, segment_pm_max, on = 'hqta_segment_id'))
        segment_peak_service['hq_transit_corr'] = segment_peak_service.apply(
                                    lambda x: x.am_max_trips > 4 and x.pm_max_trips > 4, axis = 1)
    except: ## append when all segments only have am or pm trips, not an hqta by definition
        segment_peak_service = segment_am_max.append(segment_pm_max)
        segment_peak_service['hq_transit_corr'] = False
    ##
    
    segment_peak_service = segment_peak_service.reset_index(drop=True)
    segment_peak_service['last_seg_hqta'] = segment_peak_service['hq_transit_corr'].shift(1)
    segment_peak_service['next_seg_hqta'] = segment_peak_service['hq_transit_corr'].shift(-1)
    
    ## consider segment not HQTA if both the prior and next segements are not HQTAs
    debug_dict[f'{shape_id}_sps'] = segments_with_max_stop >> inner_join(_, segment_peak_service, on = 'hqta_segment_id')
    segment_peak_service['hq_transit_corr'] = segment_peak_service.apply(
            lambda x: False if x.hq_transit_corr == False else x.last_seg_hqta or x.next_seg_hqta, axis = 1)
    segment_peak_service['hq_transit_corr'] = segment_peak_service['hq_transit_corr'].fillna(True)
    
    single_hqta = segments_with_max_stop >> inner_join(_, segment_peak_service, on = 'hqta_segment_id')
    single_hqta.drop(columns=['calitp_extracted_at', 'next_seg_hqta', 'last_seg_hqta',
                               'index', 'n_trips'], inplace=True)
    
    if shape_id in debug_ids:
        debug_dict[f'{shape_id}_single_hqta'] = single_hqta
    
    return single_hqta.reset_index(drop=True)

In [15]:
def single_operator_hqta(views):  
    
    global debug_dict
    debug_dict = {}
    
    shapes, trips, stop_times, stops = views
    
    
    distinct_routes = (trips
                   >> distinct(_.route_id, _.shape_id, _.direction_id, _keep_all = True)
                   >> select(_.calitp_itp_id, _.route_id, _.shape_id, _.direction_id, _.trip_id))
    
    route_count_by_stop = (stop_times
                         >> select(_.stop_id, _.trip_id)
                         >> inner_join(_, distinct_routes, on = "trip_id")
                         >> count(_.stop_id)
                         >> rename(n_routes = _.n)
                         >> arrange(-_.n_routes))    
    
    hqta = gpd.GeoDataFrame()
    ## start with shapes including the highest number of trips
    shapes_sorted = trips.groupby('shape_id').count().sort_values(by='trip_id', ascending=False).index
    shapes_sorted = pd.Series(shapes_sorted)
    total_shapes = len(shapes_sorted)
    print(f'there are {total_shapes} shapes total')
    for ix, shape_id in shapes_sorted.items():
        print(f'calculating for shape_id {shape_id}')
        if ix % 25 == 0:
            print(f'progress: {ix}/{total_shapes} shapes ({round(((ix/total_shapes)*100), 2)}%)')
        try:
            result = single_shape_hqta(shapes, trips, stop_times, stops, route_count_by_stop,
                                    shape_id, hqta)
            hqta = hqta.append(result)
        except:
            print(f'unable to calculate HQTA for shape_id {shape_id}')
        try:
            hqta = hqta.set_crs(shared_utils.geography_utils.CA_NAD83Albers)
        except:
            continue
    return hqta
    # return hqta.drop(columns=['n', 'departure_hour']).reset_index(drop=True)

In [16]:
debug_ids = ['940143', '1730123']

In [17]:
lbt_views = get_operator_views(170)

/opt/conda/lib/python3.9/site-packages/siuba/sql/utils.py:52: SAWarning: Dialect bigquery:bigquery will not make use of SQL compilation caching as it does not set the 'supports_statement_cache' attribute to ``True``.  This can have significant performance implications including some performance degradations in comparison to prior SQLAlchemy versions.  Dialect maintainers should seek to set this attribute to True after appropriate development and testing for SQLAlchemy 1.4 caching support.   Alternatively, this attribute may be set to False which will disable this warning. (Background on this error at: https://sqlalche.me/e/14/cprf)


loaded bus routes
loaded trips
loaded stop times
loaded stops


In [19]:
lbt_hqta = single_operator_hqta(lbt_views)

there are 97 shapes total
calculating for shape_id 510074
progress: 0/97 shapes (0.0%)
calculating for shape_id 610105
calculating for shape_id 510075
calculating for shape_id 610106
calculating for shape_id 1210143
calculating for shape_id 1210150
calculating for shape_id 460064
calculating for shape_id 460067
no line for shape 460067
calculating for shape_id 1920184
calculating for shape_id 1920183
calculating for shape_id 1910328
calculating for shape_id 1910326
calculating for shape_id 1730123
***debug shape*** 1730123
calculating for shape_id 1730122
calculating for shape_id 220076
calculating for shape_id 220080
calculating for shape_id 1710178
calculating for shape_id 1710177
calculating for shape_id 1720090
calculating for shape_id 1720091
no line for shape 1720091
calculating for shape_id 10074
calculating for shape_id 80004
calculating for shape_id 410006
calculating for shape_id 40008
calculating for shape_id 40007
no line for shape 40007
calculating for shape_id 20014
progr

In [25]:
##TODO replace map with shared utils folium, useful mouseover

In [26]:
# map_hqta(lbt_hqta)

In [27]:
lbt_hqta.shape_id.unique()

array(['510074', '610105', '510075', '610106', '1210143', '1210150',
       '460064', '1920184', '1920183', '1910328', '1910326', '1730123',
       '1730122', '220076', '220080', '1710178', '1710177', '1720090',
       '10074', '80004', '410006', '40008', '20014', '20013', '1510067',
       '410009', '710054', '710053', '1510064', '450110', '1310131',
       '1040021', '1040022', '1010087', '910149', '1030063', '910151',
       '1030064', '1810080', '230020', '210089', '1820070', '1110083',
       '1120049', '1810081', '920120', '930159', '4050003'], dtype=object)

In [28]:
# map_hqta(lbt_hqta, 'hqta_segment_id')

In [30]:
# map_hqta(lbt_hqta >> filter(_.shape_id == '1730123'), 'stop_id')

In [69]:
## investigate PCH not being an hqta...

In [35]:
## still shows issue...
lbt_hqta >> filter(_.shape_id == '1730123', _.am_max_trips >= 4, _.hq_transit_corr == False)

calitp_itp_id                                           geometry  \
10           170  POLYGON ((177962.049 -463099.837, 177961.856 -...   
18           170  POLYGON ((171789.381 -467743.362, 171793.498 -...   
22           170  POLYGON ((168037.524 -467916.250, 168037.372 -...   
23           170  POLYGON ((167357.829 -467928.625, 167358.227 -...   
24           170  POLYGON ((167281.629 -469235.162, 167281.182 -...   

    hqta_segment_id segment_sequence shape_id stop_id  am_max_trips  \
10       3062012927               10  1730123    1272             4   
18       3092845517               18  1730123    1346             7   
22       1939974416               22  1730123    1662             7   
23         78027142               23  1730123    0054             7   
24       2596408357               24  1730123    0360            10   

    pm_max_trips  hq_transit_corr  
10             3            False  
18             7            False  
22             7            False  
23             8            False  
24            10            False

In [39]:
sps = debug_dict['1730123_sps']
sps >> filter(_.shape_id == '1730123', _.am_max_trips >= 4, _.hqta_segment_id == 3092845517)
## correct here but last/next incorrect

index calitp_itp_id shape_id calitp_extracted_at  \
18      0           170  1730123          2022-02-08   

                                             geometry segment_sequence  \
18  POLYGON ((171789.381 -467743.362, 171793.498 -...               18   

    hqta_segment_id stop_id  n_trips  am_max_trips  pm_max_trips  \
18       3092845517    1346      110             7             7   

    hq_transit_corr last_seg_hqta next_seg_hqta  
18             True         False         False

In [42]:
## demo of algorithm? debugging here?
sps['hq_transit_corr'] = sps.apply(
        lambda x: False if x.hq_transit_corr == False else x.last_seg_hqta or x.next_seg_hqta, axis = 1)
sps['hq_transit_corr'] = sps['hq_transit_corr'].fillna(True)

In [43]:
sps

index calitp_itp_id shape_id calitp_extracted_at  \
0       0           170  1730123          2022-02-08   
1       0           170  1730123          2022-02-08   
2       0           170  1730123          2022-02-08   
3       0           170  1730123          2022-02-08   
4       0           170  1730123          2022-02-08   
5       0           170  1730123          2022-02-08   
6       0           170  1730123          2022-02-08   
7       0           170  1730123          2022-02-08   
8       0           170  1730123          2022-02-08   
9       0           170  1730123          2022-02-08   
10      0           170  1730123          2022-02-08   
11      0           170  1730123          2022-02-08   
12      0           170  1730123          2022-02-08   
13      0           170  1730123          2022-02-08   
14      0           170  1730123          2022-02-08   
15      0           170  1730123          2022-02-08   
16      0           170  1730123          2022-02-08   
17      0           170  1730123          2022-02-08   
18      0           170  1730123          2022-02-08   
19      0           170  1730123          2022-02-08   
20      0           170  1730123          2022-02-08   
21      0           170  1730123          2022-02-08   
22      0           170  1730123          2022-02-08   
23      0           170  1730123          2022-02-08   
24      0           170  1730123          2022-02-08   

                                             geometry segment_sequence  \
0   POLYGON ((175350.339 -454045.931, 175349.213 -...                0   
1   POLYGON ((175836.936 -454587.721, 175836.959 -...                1   
2   POLYGON ((175798.075 -455972.107, 175798.083 -...                2   
3   POLYGON ((175827.602 -457076.012, 175827.610 -...                3   
4   POLYGON ((175854.103 -458921.667, 175854.135 -...                4   
5   POLYGON ((176737.013 -460049.998, 176737.342 -...                5   
6   POLYGON ((177991.832 -460007.766, 178010.111 -...                6   
7   POLYGON ((178469.964 -460896.154, 178469.968 -...                7   
8   POLYGON ((178396.327 -461576.519, 178395.927 -...                8   
9   POLYGON ((178429.971 -462989.509, 178430.591 -...                9   
10  POLYGON ((177962.049 -463099.837, 177961.856 -...               10   
11  POLYGON ((175919.561 -463405.079, 175916.534 -...               11   
12  POLYGON ((176035.395 -463492.872, 176035.524 -...               12   
13  POLYGON ((176057.520 -464803.915, 176057.536 -...               13   
14  POLYGON ((176069.941 -466005.283, 176073.170 -...               14   
15  POLYGON ((175997.767 -467784.889, 175997.533 -...               15   
16  POLYGON ((175352.301 -467893.007, 175352.081 -...               16   
17  POLYGON ((173784.416 -467841.455, 173785.048 -...               17   
18  POLYGON ((171789.381 -467743.362, 171793.498 -...               18   
19  POLYGON ((171187.739 -467856.064, 171187.546 -...               19   
20  POLYGON ((169935.539 -467879.393, 169923.522 -...               20   
21  POLYGON ((168696.051 -467903.956, 168680.399 -...               21   
22  POLYGON ((168037.524 -467916.250, 168037.372 -...               22   
23  POLYGON ((167357.829 -467928.625, 167358.227 -...               23   
24  POLYGON ((167281.629 -469235.162, 167281.182 -...               24   

    hqta_segment_id stop_id  n_trips  am_max_trips  pm_max_trips  \
0        2631631551    4113       69             5             5   
1        3957109289    4116       35             3             2   
2        1926620051    4119       35             3             2   
3          97702661    3033       35             3             2   
4        2612414118    4124       35             3             2   
5        3971053104    1909       34             3             3   
6        1975003018    1902       31             2             2   
7          46077724    1900       31             2             2   
8        2

## Multiple Operators

In [11]:
itp_ids = tbl.gtfs_schedule.agency() >> distinct(_.calitp_itp_id) >> collect()
itp_ids = itp_ids.calitp_itp_id

In [12]:
def multiple_operator_hqta(existing=pd.DataFrame()):
    
    hqta = pd.DataFrame()    
    for _, itp_id in itp_ids.items():
        if not existing.empty:
            if str(itp_id) in existing['calitp_itp_id'].to_list():
                print(f'{itp_id} existing')
                # break
                continue
        if int(itp_id) == 200:
            continue ## skip MTC feed to use individual operator feeds
        try:
            print(f'attempting for operator {itp_id}')
            hqta = hqta.append(single_operator_hqta(get_operator_views(itp_id)))
        except:
            print(f'failed for operator {itp_id}')
        
    return hqta

In [6]:
# all_operators = multiple_operator_hqta()
# geoparquet_gcs_export(all_operators, 'all_operators')

In [3]:
all_operators = gpd.read_parquet(f'{GCS_FILE_PATH}all_operators.parquet')

In [14]:
itp_ids = itp_ids.astype(str)

In [15]:
df = tbl.gtfs_schedule.agency() >> collect() >> distinct(_.calitp_itp_id, _keep_all = True)

In [16]:
not_ran = df[~df['calitp_itp_id'].isin(all_operators['calitp_itp_id'].astype('int64'))]

In [17]:
with pd.option_context('display.max_rows', 100):
    display(not_ran)

calitp_itp_id  calitp_url_number                             agency_id  \
2              474                  0                                 TCRTA   
3              256                  0                           Porterville   
4              279                  0                                  BART   
5              200                  0                                    BA   
24             325                  0                                    68   
25              62                  0                                    68   
27             372                  0  d139e461-175c-4156-beb0-e6a23b64d978   
36              13                  0                                  1206   
38             338                  0                                   513   
42              10                  0                                    CE   
43             246                  0                                  1000   
86             254                  0                                    20   
90              15                  0                                   897   
91              41                  0                                   897   
94             323                  0                             Metrolink   
96             349                  0                                  1642   
97             315                  0                                  1535   
106            341                  0                                  1490   
116            280                  0                                   873   
117            271                  0                             Roseville   
137            344                  0                                    34   
143              7                  0                                   509   
153            312                  0                              11214031   
165            111                  0                                   FSL   
166            265                  0                                   989   
168            121                  0                                  1664   
169             97                  0                                   484   
175            122                  0                                  1707   

                               agency_name  \
2    Tulare County Regional Transit Agency   
3                      Porterville Transit   
4                   Bay Area Rapid Transit   
5                   Bay Area Rapid Transit   
24             Stanislaus Regional Transit   
25             Stanislaus Regional Transit   
27     Yolo County Transportation District   
36                                    None   
38                     Tideline Water Taxi   
42               Altamont Corridor Express   
43                                Caltrain   
86                          Plumas Transit   
90              Angel Island Tiburon Ferry   
91              Angel Island Tiburon Ferry   
94                        Metrolink Trains   
96                         Turlock Transit   
97          Sonoma Marin Area Rail Transit   
106                                 TRACER   
116                San Francisco Bay Ferry   
117                      Roseville Transit   
137                        Trinity Transit   
143                             Spirit Bus   
153                              A11214029   
165                     Folsom Stage Lines   
166                        Blossom Express   
168       Glendora Transportation Division   
169                         Duarte Transit   
175                  Glenn Transit Service   

                                            agency_url      agency_timezone  \
2                                 https://ridetcat.org  America/Los_Angeles   
3                                http://www.ridept.com  America/Los_Angeles   
4                                https://www.bart.gov/  America/Los_Angeles   
5                                https://www.bart.gov/  Americ

## Spatially Aggregate non-HQTAs

* currently unable with new flexible peak definition

In [26]:
# non_hqta = all_operators >> filter(-_.hq_transit_corr)

# non_hqta_overlaid = non_hqta.overlay(non_hqta, how='intersection')

# non_hqta_deduplicated = non_hqta_overlaid[non_hqta_overlaid['hqta_segment_id_1'] != non_hqta_overlaid['hqta_segment_id_2']]

In [27]:
# def combine_segment_ids(row):
#     both_segments = [row.hqta_segment_id_1, row.hqta_segment_id_2]
#     both_segments.sort()
#     return str(both_segments)

In [28]:
# non_hqta_deduplicated['overlap_id'] = non_hqta_deduplicated.apply(combine_segment_ids, axis=1)

In [44]:
# more_deduplicated = non_hqta_deduplicated.drop_duplicates(subset=['overlap_id'])

In [8]:
# more_deduplicated.to_parquet('./data/bus/overlaps.parquet')

In [47]:
# ##TODO rewrite to new definition... (or just drop?)

# def sum_overlaps(row):
#     row['am_peak'] = row.am_peak_1 + row.am_peak_2
#     row['pm_peak'] = row.pm_peak_1 + row.pm_peak_2
#     row['n_trips'] = row.n_trips_1 + row.n_trips_2
#     if row.am_peak >= 4 and row.pm_peak >= 4:
#         row['hq_transit_corr'] = True
#     else:
#         row['hq_transit_corr'] = False
#     return row

In [48]:
# overlaps_summed = more_deduplicated.apply(sum_overlaps, axis = 1) 

In [49]:
# new_hq = overlaps_summed[overlaps_summed['hq_transit_corr']]

In [50]:
# new_hq = new_hq[new_hq['am_peak'] != new_hq['am_peak_1']*2] ## drop likely duplicates; can rework above to do this better

In [51]:
# new_hq = new_hq.drop(columns=['calitp_itp_id_2',
#                'hqta_segment_id_2', 'index_2', 'n_trips_2', 'segment_sequence_2',
#                'shape_id_2', 'stop_id_2', 'am_peak_2', 'pm_peak_2',
#                'hq_transit_corr_2', 'index_1', 'index_2', 'n_trips_2', 'n_trips_1',
#                 'am_peak_1', 'pm_peak_1', 'hq_transit_corr_1'])

In [52]:
# new_hq = new_hq.rename(columns = {'calitp_itp_id_1':'calitp_itp_id', 'hqta_segment_id_1':'hqta_segment_id',
#        'segment_sequence_1':'segment_sequence', 'shape_id_1':'shape_id', 'stop_id_1':'stop_id'})

In [55]:
# all_operators = all_operators.append(new_hq)

In [4]:
short_dropped = all_operators[all_operators['geometry'].area > 50*400] ##50m width * 400m segment min

In [5]:
short_dropped = short_dropped.reset_index().drop(columns=['index'])

In [6]:
# short_dropped.to_parquet('./data/bus/all_operators_cleaned_appended.parquet')

In [7]:
dissolved = short_dropped.dissolve(by=['hq_transit_corr', 'shape_id']).reset_index()

In [8]:
dissolved = dissolved[dissolved['geometry'].area > 50*3000] ##50m width * 3000m shape min

In [37]:
dissolved = gpd.read_parquet(f'{GCS_FILE_PATH}shape_hqta_dissolve.parquet')

In [38]:
map_hqta(dissolved)

Map(center=[33.596015707487744, -117.87571863402819], controls=(ZoomControl(options=['position', 'zoom_in_text…

In [14]:
geoparquet_gcs_export(dissolved, 'shape_hqta_dissolve')

/home/jovyan/data-analyses/high_quality_transit_areas/utilities.py:85: UserWarning: this is an initial implementation of Parquet/Feather file support and associated metadata.  This is tracking version 0.1.0 of the metadata specification at https://github.com/geopandas/geo-arrow-spec

This metadata specification does not yet make stability promises.  We do not yet recommend using this in a production setting unless you are able to rewrite your Parquet/Feather files.

To further ignore this warning, you can do: 
import warnings; warnings.filterwarnings('ignore', message='.*initial implementation of Parquet.*')
  gdf.to_parquet(f"{name}.parquet")


### Bay Area Map Image

![map](img/bay.png)

### Los Angeles Map Image

![map](img/la.png)

## Result

* Ran and aggregated for nearly all bus operators statewide
    * Segments not containing stops will not appear as HQTA-- may need to interpolate (e.g., freeway segments)
    * Some questionable short segments
* Algorithm may be overestimating for SFMTA

### Data Issues

* 61 County Connection has a lot of na departure times?
    * strange results after dropna
* 48 B-Line similarly choppy
* 116 Fresno Area Express has whitespace in departure times
    * ValueError: time data ' 7:04:00' does not match format '%H:%M:%S'

